File: examples/orchestrator-agent/orchestrator_agent.py
"""Minimal LAP Local orchestrator; stdout contains protocol frames only."""

from __future__ import annotations

import json
import sys
from typing import Any

AGENT_ID = "org.lap.orchestrator-agent"
EXTENSION_KEY = "io.github.lambda-harness.lap.workflow.orchestrator"
sequence = 0


def emit(
    message_type: str,
    payload: dict[str, Any],
    *,
    correlation_id: str | None = None,
    run: dict[str, Any] | None = None,
) -> None:
    """Emit one ordered LAP protocol frame to standard output.

    Args:
        message_type: LAP message type to emit.
        payload: Protocol payload associated with the message.
        correlation_id: Optional Host message identifier being answered.
        run: Optional Host-owned run identity to preserve in the frame.
    """
    global sequence
    sequence += 1
    frame: dict[str, Any] = {
        "lap": "0.1",
        "id": f"orchestrator-{sequence}",
        "producer": AGENT_ID,
        "seq": sequence,
        "type": message_type,
        "payload": payload,
    }
    if correlation_id:
        frame["correlation_id"] = correlation_id
    if run:
        frame["run"] = run
    print(json.dumps(frame, ensure_ascii=False, separators=(",", ":")), flush=True)


def dispatch_from_context(payload: dict[str, Any]) -> dict[str, Any]:
    """Build one proposal from the Host-scoped workflow context.

    Args:
        payload: ``run.start`` payload containing the workflow extension.

    Returns:
        A dispatch proposal limited to the first Host-granted target.

    Raises:
        ValueError: If the required Host workflow extension is invalid.
    """
    context = payload.get("context")
    extensions = context.get("extensions") if isinstance(context, dict) else None
    extension = extensions.get(EXTENSION_KEY) if isinstance(extensions, dict) else None
    if not isinstance(extension, dict) or extension.get("version") != "0.2":
        raise ValueError("Missing LAP workflow orchestrator context.")
    allowed = extension.get("allowed_dispatches")
    if not isinstance(allowed, list) or not allowed:
        raise ValueError("Workflow orchestrator context has no dispatch targets.")
    target = allowed[0]
    if not isinstance(target, dict):
        raise ValueError(
            "Workflow orchestrator context has an invalid dispatch target."
        )
    agent_id = target.get("agent_id")
    capabilities = target.get("capabilities")
    if (
        not isinstance(agent_id, str)
        or not isinstance(capabilities, list)
        or not capabilities
    ):
        raise ValueError(
            "Workflow orchestrator context has an invalid dispatch target."
        )
    capability = capabilities[0]
    if not isinstance(capability, str):
        raise ValueError("Workflow orchestrator context has an invalid capability.")
    return {
        "dispatch": [
            {
                "agent_id": agent_id,
                "capability": capability,
                "input": payload.get("input"),
            }
        ],
    }


def handle(frame: dict[str, Any]) -> bool:
    """Handle one Host frame and report whether the Agent should keep running.

    Args:
        frame: Parsed LAP envelope supplied by the Host.

    Returns:
        ``False`` after a shutdown request; otherwise ``True``.
    """
    message_type = frame.get("type")
    if message_type == "agent.hello":
        emit(
            "agent.welcome",
            {
                "selected_lap": "0.1",
                "profiles": ["lap-local/0.1", "lap-workflow/0.2"],
                "agent_id": AGENT_ID,
                "version": "0.1.0",
                "max_concurrency": 1,
            },
            correlation_id=str(frame.get("id") or ""),
        )
        return True
    if message_type == "run.start":
        run = frame.get("run") or {}
        payload = frame.get("payload") or {}
        emit(
            "run.accepted",
            {"capability": payload.get("capability")},
            correlation_id=str(frame.get("id") or ""),
            run=run,
        )
        try:
            output = dispatch_from_context(payload)
        except ValueError as exc:
            emit(
                "run.result",
                {
                    "status": "failed",
                    "summary": "Workflow context was rejected.",
                    "error": {
                        "code": "LAP-201",
                        "message": str(exc),
                        "retryable": False,
                    },
                },
                run=run,
            )
            return True
        emit(
            "run.progress",
            {
                "phase": "planning",
                "message": "Constructed a proposal from the Host-scoped dispatch context.",
            },
            run=run,
        )
        emit(
            "run.result",
            {
                "status": "succeeded",
                "summary": "Proposed one Host-constrained dispatch.",
                "output": output,
            },
            run=run,
        )
        return True
    if message_type == "run.cancel":
        emit(
            "run.result",
            {
                "status": "cancelled",
                "summary": "Run cancelled.",
                "error": {
                    "code": "LAP-401",
                    "message": "Cancellation requested.",
                    "retryable": False,
                },
            },
            run=frame.get("run") or {},
        )
        return True
    if message_type == "agent.shutdown":
        return False
    return True


for raw in sys.stdin:
    try:
        message = json.loads(raw)
        if not isinstance(message, dict):
            raise ValueError("frame must be an object")
    except (json.JSONDecodeError, ValueError):
        sys.exit(2)
    if not handle(message):
        break

File: examples/orchestrator-agent-node/orchestrator_agent.js
// LAP Node.js external orchestrator. Stdout is protocol-only.
"use strict";

const readline = require("node:readline");

const AGENT_ID = "org.lap.orchestrator-agent-node";
const AGENT_VERSION = "0.1.0";
const EXTENSION_KEY = "io.github.lambda-harness.lap.workflow.orchestrator";
const AGENT_ID_PATTERN = /^[a-z][a-z0-9.-]{2,127}$/;
const CAPABILITY_PATTERN = /^[a-z][a-z0-9._-]{2,127}$/;
let sequence = 0;

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function asObject(value) {
  return isObject(value) ? value : {};
}

function hasExactKeys(value, keys) {
  if (!isObject(value)) return false;
  const actual = Object.keys(value).sort();
  return actual.length === keys.length && actual.every((key, index) => key === keys[index]);
}

function emit(messageType, payload, correlationId = "", run = null) {
  sequence += 1;
  const frame = {
    lap: "0.1",
    id: `node-orchestrator-${sequence}`,
    producer: AGENT_ID,
    seq: sequence,
    type: messageType,
    payload,
  };
  if (correlationId) frame.correlation_id = correlationId;
  if (isObject(run)) frame.run = run;
  process.stdout.write(`${JSON.stringify(frame)}\n`);
}

function dispatchFromContext(payload) {
  const context = asObject(payload.context);
  const extensions = asObject(context.extensions);
  const extension = extensions[EXTENSION_KEY];
  if (!hasExactKeys(extension, ["allowed_dispatches", "version"]) || extension.version !== "0.2") {
    throw new Error("invalid context");
  }
  const allowed = extension.allowed_dispatches;
  if (!Array.isArray(allowed) || allowed.length === 0) throw new Error("invalid context");

  let previousAgentId = "";
  let selectedAgentId = "";
  let selectedCapability = "";
  for (let index = 0; index < allowed.length; index += 1) {
    const target = allowed[index];
    if (!hasExactKeys(target, ["agent_id", "capabilities"])) throw new Error("invalid context");
    const agentId = target.agent_id;
    const capabilities = target.capabilities;
    if (typeof agentId !== "string" || !AGENT_ID_PATTERN.test(agentId)) throw new Error("invalid context");
    if (previousAgentId && agentId <= previousAgentId) throw new Error("invalid context");
    if (!Array.isArray(capabilities) || capabilities.length === 0) throw new Error("invalid context");

    let previousCapability = "";
    for (const capability of capabilities) {
      if (typeof capability !== "string" || !CAPABILITY_PATTERN.test(capability)) {
        throw new Error("invalid context");
      }
      if (previousCapability && capability <= previousCapability) throw new Error("invalid context");
      previousCapability = capability;
    }
    if (index === 0) {
      selectedAgentId = agentId;
      selectedCapability = capabilities[0];
    }
    previousAgentId = agentId;
  }

  return {
    dispatch: [{
      agent_id: selectedAgentId,
      capability: selectedCapability,
      input: Object.prototype.hasOwnProperty.call(payload, "input") ? payload.input : {},
    }],
  };
}

function handle(frame) {
  const messageType = frame.type;
  if (messageType === "agent.hello") {
    emit("agent.welcome", {
      selected_lap: "0.1",
      profiles: ["lap-local/0.1", "lap-workflow/0.2"],
      agent_id: AGENT_ID,
      version: AGENT_VERSION,
      max_concurrency: 1,
    }, String(frame.id || ""));
    return true;
  }
  if (messageType === "run.start") {
    const run = asObject(frame.run);
    const payload = asObject(frame.payload);
    emit("run.accepted", { capability: payload.capability }, String(frame.id || ""), run);
    let output;
    try {
      output = dispatchFromContext(payload);
    } catch (_error) {
      emit("run.result", {
        status: "failed",
        summary: "Workflow context was rejected.",
        error: {
          code: "LAP-201",
          message: "A valid Host-scoped workflow orchestrator context is required.",
          retryable: false,
        },
      }, "", run);
      return true;
    }
    emit("run.progress", {
      phase: "planning",
      message: "Constructed a proposal from the Host-scoped dispatch context.",
    }, "", run);
    emit("run.result", {
      status: "succeeded",
      summary: "Proposed one Host-constrained dispatch.",
      output,
    }, "", run);
    return true;
  }
  if (messageType === "run.cancel") {
    emit("run.result", {
      status: "cancelled",
      summary: "Run cancelled.",
      error: { code: "LAP-401", message: "Cancellation requested.", retryable: false },
    }, "", asObject(frame.run));
    return true;
  }
  return messageType !== "agent.shutdown";
}

const input = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
let stopped = false;
input.on("line", (raw) => {
  if (stopped) return;
  let frame;
  try {
    frame = JSON.parse(raw);
    if (!isObject(frame)) throw new Error("frame must be an object");
  } catch (_error) {
    process.stderr.write("invalid LAP frame\n");
    process.exitCode = 2;
    stopped = true;
    input.close();
    return;
  }
  if (!handle(frame)) {
    stopped = true;
    input.close();
  }
});
